src/query/subquery_cache.c

```diff
@@ -0,0 +1,724 @@
+  key = sq_make_key (xasl);
+  if (key == NULL)
+    {
+      XASL_SET_FLAG (xasl, XASL_SQ_CACHE_NOT_CACHING);
+      return false;
+    }
+  if (!XASL_IS_FLAGED (xasl, XASL_SQ_CACHE_INITIALIZED))
+    {
+      sq_cache_initialize (xasl);
+      xasl->sq_cache.stats.miss++;
+      sq_free_key (key);
+      return false;
+    }
+
+  ret = (sq_val *) mht_get (xasl->sq_cache.ht, key);
+  if (ret == NULL)
+    {
+      xasl->sq_cache.stats.miss++;
+      sq_free_key (key);
+      return false;
+    }
+
+  sq_unpack_val (ret, regu_var);
+  sq_free_key (key);
+
+  xasl->sq_cache.stats.hit++;
+  return true;
+}
+
+/*
+ * sq_cache_destroy () - Destroys the cache for a given XASL node.
+ *   xasl(in): The XASL node for which the cache is being destroyed.
+ *
+ * This function destroys the cache associated with a given XASL node. It clears all cache entries and then destroys the hash
+ * table itself. It also resets cache-related flags and counters for the XASL node. This function is called when a XASL node is
+ * no longer needed or before it is deallocated.
+ */
+void
+sq_cache_destroy (xasl_node * xasl)
+{
+  if (XASL_IS_FLAGED (xasl, XASL_SQ_CACHE_INITIALIZED))
+    {
+      er_log_debug (ARG_FILE_LINE,
+		    "destroy sq_cache at xasl %p\ncache info : \n\thit : %10lu\n\tmiss: %10lu\n\tsize: %10lu Bytes\n",
+		    xasl, xasl->sq_cache.stats.hit, xasl->sq_cache.stats.miss, xasl->sq_cache.size);
+      mht_clear (xasl->sq_cache.ht, sq_rem_func, NULL);
+      mht_destroy (xasl->sq_cache.ht);
+    }
+  XASL_CLEAR_FLAG (xasl,
+		   XASL_SQ_CACHE_ENABLED | XASL_SQ_CACHE_INITIALIZED | XASL_SQ_CACHE_NOT_CACHING |
+		   XASL_SQ_CACHE_NOT_CACHING_CHECKED);
+  xasl->sq_cache.ht = NULL;
+}
+
+/*
+ * sq_check_enable () - Checks if caching is enabled for a given XASL node and updates the cache status flags accordingly.
+ *   xasl(in): The XASL node to check and update cache status for.
+ *
+ * This function determines whether caching is enabled for a specified XASL node. It first checks if the node is null, 
+ * if caching has been explicitly disabled, or if the node's status is neither CLEARED nor INITIALIZED, returning FALSE in 
+ * any of these cases. 
+ * If the XASL_SQ_CACHE_ENABLED is already set, it returns TRUE, indicating caching is enabled. 
+ * Otherwise, it sets the XASL_SQ_CACHE_ENABLED if not already set, checks if the node should not be cached by calling 
+ * sq_walk_xasl_check_not_caching(), and updates the flag accordingly. The function returns FALSE if caching is not enabled 
+ * by the end of its execution. 
+ * This ensures that the node's caching status is accurately reflected and updated based on its current state and any conditions 
+ * that might prevent caching.
+ */
+int
+sq_check_enable (xasl_node * xasl)
+{
+  if (!(xasl) || XASL_IS_FLAGED (xasl, XASL_SQ_CACHE_NOT_CACHING)
+      || !(xasl->status == XASL_CLEARED || xasl->status == XASL_INITIALIZED))
+    {
+      return FALSE;
+    }
+  if (XASL_IS_FLAGED (xasl, XASL_SQ_CACHE_ENABLED))
```

Comment:
Is XASL_SQ_CACHE_ENABLED necessary?
I think it can be checked with XASL_SQ_CACHE_NOT_CACHING_CHECKED and XASL_SQ_CACHE_NOT_CACHING.